// liveramp-ui-toolkit/js/ui_toolkit.js
require('../sass/ui_toolkit.sass');

module.exports = {
  UiButtonGroup: require('./components/uiButtonGroup.jsx').default,
  UiCheckbox: require('./components/uiCheckbox.jsx').default,
  UiFilterGroup: require('./components/uiFilterGroup.jsx').default,
  UiHeader: require('./components/uiHeader.jsx').default,
  UiIcon: require('./components/uiIcon.jsx').default,
  UiInput: require('./components/uiInput.jsx').default,
  UiLabel: require('./components/uiLabel.jsx').default,
  UiSlider: require('./components/uiSlider.jsx').default,
  UiLoadingComponent: require('./components/uiLoadingComponent.jsx').default,
  UiSelect: require('./components/uiSelect.jsx').default,
  UiIdentifierSelect: require('./components/uiIdentifierSelect.jsx').default,
  UiSearch: require('./components/uiSearch.jsx').default,
  UiTitle: require('./components/uiTitle.jsx').default,
  UiTabbedView: require('./components/uiTabbedView.jsx').default,
  UiSubHeader: require('./components/uiSubHeader.jsx').default,
  UiProgressMeter: require('./components/uiProgressMeter.jsx').default,
  UiRadioButtonGroup: require('./components/uiRadioButtonGroup.jsx').default,
  UiRadioButton: require('./components/uiRadioButton.jsx').default,
  UiTile: require('./components/uiTile.jsx').default,
  UiWizardControls: require('./components/uiWizardControls.jsx').default,
  UiDropdown: require('./components/uiDropdown.jsx').default,
  UiCollapse: require('./components/uiCollapse.jsx').default,
  UiAlert: require('./components/uiAlert.jsx').default,
  UiFileUploadStatus: require('./components/uiFileUploadStatus.jsx').default,
  UiSwitch: require('./components/uiSwitch.jsx').default,
  UiTable: require('./components/uiTable.jsx').default,
  UiSplitButton: require('./components/uiSplitButton.jsx').default,
  UiWizard: require('./components/uiWizard.jsx').default
};
