// liveramp-ui-toolkit/js/components/uiCheckbox.jsx
const docs = {
  title: "UiCheckbox",
  propDetails: [
    { title: "value",
      required: true,
      type: "integer",
      description: "The value for the checkbox." },
    { title: "checked",
      required: true,
      type: "boolean",
      description: "Whether or not the checkbox is checked." },
    { title: "handleChange",
      required: true,
      type: "function",
      description: "Function that updates state in the parent component - takes in the value and the current state of the checkbox." },
    { title: "label",
      required: false,
      type: "string",
      description: "A label for the checkbox." },
    { title: "disabled",
      required: false,
      type: "boolean",
      description: "Whether or not the checkbox is disabled." },
    { title: "color",
      required: false,
      type: "string",
      description: "If 'grey', checkbox will be grey." }
  ]
}

export default function UiCheckbox(props) {
  // User provides:
  //  - value for the checkbox (value) : integer
  //  - a boolean to determine if the checkbox is checked (checked) : boolean
  //  - change handling function that takes in the value, and the new state of the checkbox as an argument (handleChange) : function
  //  - optional label (label) : string
  //  - optional boolean to disable the checkbox (disabled) : boolean

  const { value, checked, handleChange, label, disabled, color } = props;

  const handleClick = (e) => {
    e.stopPropagation();
    if (!disabled) {
      handleChange(value, !checked);
    }
  };

  var checkboxClassName = "lr-ui-checkbox-wrapper";

  if (checked) {
    checkboxClassName += " checked";
  };

  if (disabled) {
    checkboxClassName += " disabled";
  };

  if (color === 'grey') {
    checkboxClassName += " grey"
  }


  return (
    <div className="lr-ui-checkbox-row">
      <div onClick={handleClick} className="lr-ui-checkbox-container">
        <input
          className="lr-ui-checkbox"
          type="checkbox"
          value={value}
          defaultChecked={checked}
          disabled={disabled === true}
        />
        <div className={checkboxClassName} />
        { label ? <div className={ cn("lr-ui-checkbox-label", {'grey': (color === 'grey') } ) }>{label}</div> : null }
      </div>
    </div>
  );
};

export { docs };
