// liveramp-ui-toolkit/js/components/uiCollapse.jsx
/*
 * example:
 *      const blurb = <p>Nice to blurb you</p>;
 *      return <UiCollapse
 *               expanded={true}
 *               expandPrompt={<span className='spanny'>Show Details</span>}
 *               collapsePrompt='Hide Details'
 *               info={blurb} />
 */

 const docs = {
   title: "UiCollapse",
   propDetails: [
     { title: "expanded",
       required: true,
       type: "boolean",
       description: "Initial expanded state of collapse." },
     { title: "expandPrompt",
       required: true,
       type: "string",
       description: "The prompt when the collapse is expanded." },
     { title: "collapsePrompt",
       required: true,
       type: "string",
       description: "The prompt when the collapse is collapsed." },
     { title: "info",
       required: true,
       type: "string or component",
       description: "The information to show when expanded." },
     { title: "expandUp",
       required: false,
       type: "boolean",
       description: "should the info expand upwards? defaults to expand below prompt"
     }
   ]
 }
const UiCollapse = React.createClass({
  getInitialState () {
    return {
      expanded: !!this.props.expanded
    };
  },
  render () {
    const {expandPrompt, collapsePrompt, info, expandUp=false} = this.props;
    const {expanded} = this.state;
    const prompt = (
      <div className='lr-ui-collapse-prompt' onClick={this.toggleExpanded}>
          {expanded ? collapsePrompt : expandPrompt}
      </div>
    );
    return (
      <div className='lr-ui-collapse'>
        {!expandUp ? prompt : ''}
        {expanded && info}
        {expandUp ? prompt : ''}
      </div>
      );
  },
  toggleExpanded () {
    this.setState({expanded: !this.state.expanded});
  }
});

export default UiCollapse;

export { docs };
