// liveramp-ui-toolkit/js/components/uiHeader.jsx
const docs = {
  title: "UiHeader",
  propDetails: [
    { title: "textTitle",
      required: true,
      type: "string",
      description: "The title for the header." },
    { title: "subHeader",
      required: false,
      type: "string",
      description: "An optional text or component placed to the right of the header." },
    { title: "rightSideContent",
      required: false,
      type: "component",
      description: "Content to go on the right side of the header." },
    { title: "helpText",
      required: false,
      type: "string",
      description: "Help text to go underneath the header." }
  ]
}

export default function UiHeader(props) {
  // User provides:
  //  - a title for the header (textTitle) : string
  //  - an optional text or component to go to the right of the header (subHeader) : string or component
  //  - optional text or content to go on the right side of the header (rightSideContent) : string or component
  //  - optional help text underneath the header (helpText) : string

  const { textTitle, subHeader, rightSideContent, helpText } = props;

  return (
    <div className="lr-ui-header">
      <h1 className="lr-ui-header-text-title">{textTitle}
        {subHeader ? <div className="lr-ui-header-subheader">{subHeader}</div> : null}
        {rightSideContent ? <div className="lr-ui-header-right-side-content">{rightSideContent}</div> : null}
      </h1>
      { helpText !== undefined ? <div className="lr-ui-header-help-text">{helpText}</div> : null}
    </div>
  );
};

export { docs };
