// liveramp-ui-toolkit/js/components/uiIdentifierSelect.jsx
import UiIcon from './uiIcon.jsx';
import ReactTooltip from 'react-tooltip';

const docs = {
  title: "UiIdentifierSelect",
  propDetails: [
    { title: "title",
      required: true,
      type: "string",
      description: "A title for the UiIdentifierSelect.",
    },
    { title: "identifiers",
      required: true,
      type: "array",
      description: "An array of objects, or single object, each of which has a checked (boolean), disabled(boolean), value, and icon (UiIcon).",
    },
    { title: "handleChange",
      required: undefined,
      type: "function",
      description: "A function that takes in the value of each identifier in identifiers to handle the user selecting or deselecting an identifier. Required unless displayOnly is true.",
    },
    { title: "displayOnly",
      required: false,
      type: "boolean",
      description: "If true, the icons will be unclickable, and their color will be static.",
    }
  ]
};

export default function UiIdentifierSelect (props) {
  let { title, handleChange, identifiers, displayOnly=false } = props;

  if (!Array.isArray(identifiers)) {
    identifiers = [identifiers];
  }

  return (
    <div className={cn('lr-ui-identifier-select-container', {'displayOnly': displayOnly})}>
      <div className="lr-ui-identifier-select-label">{title}</div>
      {
        _.map(identifiers, (identifier, index) => {
          const handleClick = () => {
            if (!identifier["disabled"]) {
              handleChange(identifier["value"], !identifier["checked"])
            }
          };

          if(displayOnly) {
            var color = identifier["checked"] ? "base-green" : "medium-gray";
          } else {
            color = "white"
          }

          return (
            <div
              className={cn('lr-ui-identifier-select-button', {checked: identifier["checked"], disabled: identifier["disabled"]})}
              onClick={displayOnly ? x => x : handleClick}
              key={index}
              data-tip={identifier["value"]}
              data-effect='solid'
            >
              <input
                className="lr-ui-identifier-select-input"
                type="checkbox"
                defaultChecked={identifier["checked"]}
                disabled={identifier["disabled"] === true}
              />
              <UiIcon
                icon={identifier["icon"]}
                dimensions={[32, 32]}
                color={color}
              />
            </div>
          );
        })
      }
      <ReactTooltip />
    </div>
  );
};

export { docs };
