// liveramp-ui-toolkit/js/components/uiRadioButton.jsx
const docs = {
  title: "UiRadioButton",
  propDetails: [
    { title: "label",
      required: true,
      type: "string",
      description: "Label." },
    { title: "value",
      required: true,
      type: "integer",
      description: "A value for the checkbox." },
    { title: "handleChange",
      required: true,
      type: "function",
      description: "Function to handle click - takes the value as an input." },
    { title: "selected",
      required: true,
      type: "boolean",
      description: "Whether or not the checkbox is selected." },
    { title: "disabled",
      required: false,
      type: "boolean",
      description: "Whether or not the checkbox is disabled." }
  ]
}

export default function UiRadioButton(props) {
  /* User provides:
   * * a label for the radio button (label) : string
   * * the selected status of the radio button (selected) : boolean
   * * a value that will get propagated through the handler when selected (value) : string
   * * a handler function which handles clicks and takes the value as an input (handleChange) : function
   */
  const {label, value, selected, disabled, handleChange} = props;

  const handleClick = () => {
    if (!disabled) {
      handleChange(value);
    }
  };

  return (
    <div className="lr-ui-radio-button-container"  onClick={ handleClick }>
      <input
        className="lr-ui-radio-button-input"
        name={ label }
        type="radio"
        value={ value }
        disabled={ disabled }
        checked={ selected }
        onChange={ handleClick }
      />
      <div className="lr-ui-radio-button"></div>
      <div className="lr-ui-radio-button-label">
        { label }
      </div>
    </div>
  );
};

export { docs };
