// liveramp-ui-toolkit/js/components/uiSearch.jsx
import UiIcon from './uiIcon.jsx';

const docs = {
  title: "UiSearch",
  propDetails: [
    { title: "inputValue",
      required: true,
      type: "string",
      description: "The value of the text field." },
    { title: "placeholder",
      required: true,
      type: "string",
      description: "A placeholder value for the field when the inputValue is empty." },
    { title: "handleChange",
      required: true,
      type: "function",
      description: "Updates state in the parent component, takes in the change event." },
    { title: "customStyle",
      required: false,
      type: "object",
      description: "A custom style object to be applied to the containing div." },
    { title: "handleEnter",
      required: false,
      type: "function",
      description: "A function to be called when the user presses `enter`." },
    { title: "hideBorder",
      required: false,
      type: "boolean",
      description: "An optional boolean to hide the bottom border." }
  ]
}


export default function UiSearch(props) {
  // User must pass in:
  //  - the value of the text field (inputValue) : string
  //  - custom inline styles for the outer div (customStyle) : object
  //  - placeholder value for the input field when it's empty (placeholder) : string
  //  - a function to update state in the parent component that takes in the change event (handleChange) : function
  //  - an optional function that handles the user pressing 'enter' in the search field (handleEnter) : function
  //  - an optional boolean to hide the bottom border (hideBorder) : boolean

  const { inputValue, customStyle, placeholder, handleChange, handleEnter, hideBorder } = props;

  const keyDown = function(e) {
    if (e.keyCode === 13 || e.which === 13) {
      if(handleEnter) {
        handleEnter(e);
      }
    }
  }

  return (
    <div style={customStyle} className="lr-ui-search">
      <input
        type="text"
        className={ hideBorder ? "no-border" : "" }
        value={inputValue}
        placeholder={placeholder}
        onChange={handleChange}
        onKeyDown={keyDown}
        />
      <UiIcon
        icon={'search'}
        dimensions={[13, 13]}
      />
    </div>
  );
};

export { docs };
