// liveramp-ui-toolkit/js/components/uiTitle.jsx
import moment from 'moment';
import UiSwitch from '../components/uiSwitch.jsx';

const docs = {
  title: "UiTitle",
  propDetails: [
    { title: "textTitle",
      required: true,
      type: "string",
      description: "The title to be displayed." },
    { title: "date",
      required: false,
      type: "date",
      description: "An optional date to display as a subtitle to the title." },
    { title: "showSwitch",
      required: false,
      type: "boolean",
      description: "Whether or not to show an optional switch." },
    { title: "switchOn",
      required: false,
      type: "boolean",
      description: "Whether or not the switch is on." },
    { title: "handleChange",
      required: false,
      type: "function",
      description: "A function to update the state of switchOn in the parent container." }
  ]
}

export default function UiTitle(props) {
  // user should pass in:
  // - a title (textTitle)
  // - an optional date (date)
  // - whether or not they want to include an optional masterSwitch (showSwitch)
  // - a boolean, controlled by the parent, that says whether the switch is on or off (switchOn)
  // - a function that changes the state of switchOn (handleChange)

  const { textTitle, date, showSwitch, switchOn, handleChange } = props;

  const getDate = () => {
    return (
      <div className="lr-ui-title-date">{moment(date).format('L')}</div>
    );
  };

  const getSwitch = () => {
    return (
    <UiSwitch
      switchOn={switchOn}
      handleChange={handleChange}
      />
    );
  };

  return (
    <div className="lr-ui-title">
      <div className="lr-ui-title-text-title">{textTitle}</div>
      {date ? getDate() : null}
      {showSwitch ? getSwitch() : null}
    </div>
  );
};


export { docs };
