// liveramp-ui-toolkit/js/sandboxComponents/uiCheckboxSandbox.jsx
import UiCheckbox from '../components/uiCheckbox.jsx';



const CheckboxSandbox = React.createClass({
  getInitialState () {
    return ({
      firstChecked: false,
      secondChecked: true,
      thirdChecked: false
    });
  },

  render() {
    return (
      <div>
        <UiCheckbox
          label={"Not Disabled"}
          value={123}
          checked={this.state.firstChecked}
          handleChange={this.toggleChecked}
          />
        <UiCheckbox
          label={"Not Disabled"}
          value={234}
          checked={this.state.secondChecked}
          handleChange={this.toggleSecondChecked}
          />
        <UiCheckbox
          label={"Not Disabled"}
          value={345}
          checked={this.state.thirdChecked}
          handleChange={this.toggleThirdChecked}
          />
        <br/><br/><br/><br/><br/>
        <UiCheckbox
          label={"Disabled and unchecked"}
          disabled={true}
          checked={false}
          />
        <br/><br/><br/><br/><br/>
        <UiCheckbox
          label={"Disabled and checked"}
          disabled={true}
          checked={true}
          />
      </div>

    );
  },

  toggleChecked(value) {
    console.log(value);
    this.setState({
      firstChecked: !this.state.firstChecked
    });
  },
  toggleSecondChecked(value) {
    console.log(value);
    this.setState({
      secondChecked: !this.state.secondChecked
    });
  },
  toggleThirdChecked(value) {
    console.log(value);
    this.setState({
      thirdChecked: !this.state.thirdChecked
    })
  },

});

export default CheckboxSandbox;
