// liveramp-ui-toolkit/js/sandboxComponents/uiIdentifierSelectSandbox.jsx
import UiIdentifierSelect from '../components/uiIdentifierSelect.jsx';

const UiIdentifierSelectSandbox = React.createClass({
  getInitialState() {
    return {
      includedIdentifiers: ["apple", "android", "user"],
      disabledIdentifiers: ["postal", "user"]
    }
  },

  render() {
    const identifiers = [
      {
        checked: this.state.includedIdentifiers.includes("apple"),
        disabled: this.state.disabledIdentifiers.includes("apple"),
        value: "apple",
        icon: "apple"
      },
      {
        checked: this.state.includedIdentifiers.includes("android"),
        disabled: this.state.disabledIdentifiers.includes("android"),
        value: "android",
        icon: "android"
      },
      {
        checked: this.state.includedIdentifiers.includes("user"),
        disabled: this.state.disabledIdentifiers.includes("user"),
        value: "user",
        icon: "user"
      },
      {
        checked: this.state.includedIdentifiers.includes("postal"),
        disabled: this.state.disabledIdentifiers.includes("postal"),
        value: "postal",
        icon: "postal"
      }
    ];

    const divStyles = {
      padding: '20px'
    }

    return (
      <div style={divStyles}>
        <UiIdentifierSelect
          title="Skyler's Test Identifiers"
          handleChange={this.updateIdentifiers}
          identifiers={identifiers}
          displayOnly={false}
        />
      </div>
    );
  },

  updateIdentifiers(identifier) {
    let includedIdentifiers = this.state.includedIdentifiers;

    if (this.state.includedIdentifiers.includes(identifier)) {
      includedIdentifiers = _.without(includedIdentifiers, identifier);
    } else {
      includedIdentifiers.push(identifier);
    }

    this.setState({ includedIdentifiers: includedIdentifiers });
  }
});

export default UiIdentifierSelectSandbox;
