// liveramp-ui-toolkit/js/sandboxComponents/uiInputSandbox.jsx
import UiInput from '../components/uiInput.jsx';

const UiInputSandbox = React.createClass({
  getInitialState() {
    return ({
      inputValueFirst: "",
      inputValueLast: "",
      inputValuePassword: "",
      valid: true,
      populated: false,
      placeholder: "Hello",
      disabled: true
    })
  },

  render() {
    return (
      <div>
        <UiInput
          inputValue={this.state.inputValueFirst}
          label={"First Name"}
          placeholder={this.state.placeholder}
          valid={ !this.state.inputValueFirst || this.state.inputValueFirst.length > 3 }
          labelShift={true}
          handleChange={ (e) => { this.setState({inputValueFirst: e.target.value}); } }
          disabled={this.state.disabled}
        />
        <UiInput
          isPassword={false}
          inputValue={this.state.inputValueLast}
          label={"Last Name"}
          placeholder={this.state.placeholder}
          valid={ !this.state.inputValueLast || this.state.inputValueLast.length > 3 }
          labelShift={false}
          tooltipOnOverflow={true}
          tooltipText={this.state.inputValueLast}
          tooltipId={'mytooltipid2232'}
          handleChange={ (e) => { this.setState({inputValueLast: e.target.value}); } }
        />
        <UiInput
          isPassword={true}
          inputValue={this.state.inputValuePassword}
          label={"Password"}
          placeholder={this.state.placeholder}
          valid={true}
          labelShift={false}
          handleChange={ (e) => { this.setState({inputValuePassword: e.target.value}); } }
        />
      </div>
    )
  }

});

export default UiInputSandbox;
