// liveramp-ui-toolkit/js/sandboxComponents/uiLabelSandbox.jsx
import UiLabel from '../components/uiLabel.jsx';

// user should pass in:
// - a label (label) : string
// - a boolean to determine whether or not there should be a switch (showSwitch) : boolean
// - whether switch should default to on or off (switchOn) : boolean
// - optional boolean for whether or not a select should be shown (showSelect) : boolean
// - options for the select field, in a similar format as the uiSelect inputs in uiFilterGroup (labelSelectOptions) : object
//      - name: the filter's unique name, to be used as placeholder text
//      - selectOptions: a list of option hashes, formatted like [ { label: 'Option1', value: 1 }, { ... } ]
//      - isMulti: a boolean that determines if the input is a multi-select
//      - selected: a value, or list of values, that are selected
//      - allowCreate: an optional boolean which determines if the user can create options within the group
// - a function that handles changes to the select field in the parent component (handleSelectChange) : function
// - a function that changes the switch state in the parent component (handleSwitchChange) : function

const selectOptions = {
  name: "Day",
  selectOptions: [{label: "Day", value: 1}, {label: "Week", value: 2}, {label: "Month", value: 3}],
  isMulti: false,
  allowCreate: false
};

const UiLabelSandbox = React.createClass({
  getInitialState() {
    return (
      {
        selectedFrequency: 1,
        showNotifications: true
      }
    )
  },

  render() {
    return (
      <UiLabel
        label={"Show Notifications Every "}
        showSwitch={true}
        switchOn={this.state.showNotifications}
        showSelect={true}
        labelSelectOptions={_.extend(selectOptions, {selected: this.state.selectedFrequency})}
        handleSelectChange={this.handleSelectChange}
        handleSwitchChange={this.handleSwitchChange}
        />
    );
  },

  handleSelectChange(value) {
    this.setState({
      selectedFrequency: value
    });
  },
  handleSwitchChange() {
    this.setState({
      showNotifications: !this.state.showNotifications
    });
  }
});

export default UiLabelSandbox;
