// liveramp-ui-toolkit/js/sandboxComponents/uiSelectSandbox.jsx
import UiSelect from '../components/uiSelect.jsx';

const selectOptions = [
  {
    label: "Hillary Clinton",
    value: 1
  },
  {
    label: "Donald Trump",
    value: 2
  },
  {
    label: "Pardha Ponugoti",
    value: 3
  }
];

const filterGroupList = [
  {
    name: "Continents",
    selectOptions: [
      {
        label: "North America",
        value: 1
      },
      {
        label: "South America",
        value: 2
      },
      {
        label: "Africa",
        value: 3
      },
      {
        label: "Australia",
        value: 4
      },
      {
        label: "Asia",
        value: 5
      },
      {
        label: "Antarctica",
        value: 6
      },
      {
        label: "Europe",
        value: 7
      }
    ],
    isMulti: true
  },
  {
    name: "Apex Engineers",
    selectOptions: [
      {
        label: "Parker",
        value: 1
      },
      {
        label: "Alfonso",
        value: 2
      },
      {
        label: "Tevy",
        value: 3
      },
      {
        label: "Pardha",
        value: 4
      },
      {
        label: "Uncle Charles",
        value: 5
      },
      {
        label: "Young Charles",
        value: 6
      },
      {
        label: "Josh",
        value: 7
      }
    ],
    isMulti: false
  }
];


const UiSelectSandbox = React.createClass({
  getInitialState() {
    return {
      selected: null
    }
  },

  render() {
    return (
      <div>
        <h3>ALLOWING CREATION</h3>
        <UiSelect
          label={"POTUS"}
          selected={this.state.selected}
          selectOptions={selectOptions}
          placeholder={"POTUS"}
          isMulti={true}
          arrowOnHover={true}
          allowCreate={true}
          handleChange={this.handleSelectChange}
          />
        <button onClick={this.buttonClick}> SUBMIT </button>
      </div>
    );
  },

  handleSelectChange(value) {
    if (value === "") {
      this.setState({
        selected: []
      });
    } else if (this.state.selected !== null) {
      this.setState({
        selected: _.map(value.split(','), (el) => {
          if (isNaN(el)) {
            const newValue = selectOptions[selectOptions.length-1].value + 1;
            selectOptions.push({
              label: el,
              value: newValue
            });
            return newValue;
          } else {
            return parseInt(el);
          }
        })
      });
    } else {
      if (isNaN(value)) {
        const newValue = selectOptions[selectOptions.length-1].value + 1;
        selectOptions.push({
          label: value,
          value: newValue
        });
        this.setState({
          selected: [newValue]
        });
      } else {
        this.setState({
          selected: [value]
        })
      }
    }

  },

  buttonClick(e) {
    e.preventDefault();

    console.log("BUTTON CLICKED");
    console.log(this.state);
    console.log("------------------------");
  }

});

export default UiSelectSandbox;

// import UiSelect from '../components/uiSelect.jsx';

// var options = [
//     { value: 'one', label: 'One' },
//     { value: 'two', label: 'Two' }
// ];

// function logChange(val) {
//     console.log("Selected: " + val);
// };

// function deleteOption(e) {
//   e.stopPropagation();
//   e.preventDefault();
//   console.log("DELETE ");
// };

// function mouseUp(e) {
//   e.stopPropagation();
//   e.preventDefault();
//   console.log("Mouse up");
// };

// function optionRenderer(option) {
//   return (
//     <div>
//       <div style={{display: "inline-block", cursor: "crosshair"}}>
//         {option.value + " doe!!"}
//       </div>
//       <div style={{float: 'right', display: "inline-block", cursor: "help", width: "40px"}} onMouseDown={deleteOption} onClick={deleteOption}>
//         x
//       </div>
//     </div>
//   );
// };

// function logValueClick(value, event) {
//   console.log(event.target);
// };

// const UiSelectSandbox = React.createClass({
//   getInitialState() {
//     return ({
//       selected: null
//     });
//   },

//   render () {
//     return (
//       <div>
//         <UiSelect
//           label={"Testing out deletion of options"} />
//         <UiSelect
//           placeholder={"Cannot select"}
//           value={this.state.selected}
//           options={options}
//           optionRenderer={optionRenderer}
//           onChange={this.logChange}
//           allowCreate={true}
//           onValueClick={logValueClick}
//           />
//       </div>
//     );
//   },

//   logChange() {
//     this.setState({
//       selected: null
//     })
//   }


// });


// export default UiSelectSandbox;
