// liveramp-ui-toolkit/js/sandboxComponents/uiSwitchSandbox.jsx
import UiSwitch from '../components/uiSwitch.jsx';
import UiSelect from '../components/uiSelect.jsx';

const demBallot = [
  {
    label: "Hillary Clinton",
    value: 1
  },
  {
    label: "Bernie Sanders",
    value: 2
  },
  {
    label: "Martin O'Malley",
    value: 3
  }
];

const repubBallot = [
  {
    label: "Donald Trump",
    value: 1
  },
  {
    label: "Ted Cruz",
    value: 2
  },
  {
    label: "Marco Rubio",
    value: 3
  },
  {
    label: "Ben Carson",
    value: 4
  }
];

const UiSwitchSandbox = React.createClass({
  getInitialState() {
    return ({
      switchState: true,
      selected: null,
      selectOptions: demBallot
    })
  },

  render() {
    return (
      <div>
        <h4>Democratic ballot?</h4>
        <UiSwitch
          switchOn={this.state.switchState}
          handleChange={this.triggerAtSwitchToggle} />
        <UiSelect
          label={"Primary Ballot with Write-Ins"}
          selected={this.state.selected}
          selectOptions={this.state.selectOptions}
          placeholder={"Primary"}
          isMulti={true}
          arrowOnHover={true}
          allowCreate={true}
          handleChange={this.handleSelectChange} />
      </div>

      )
  },

  handleSelectChange(value) {
    if (value === "") {
      this.setState({
        selected: []
      });
    } else if (this.state.selected !== null) {
      this.setState({
        selected: _.map(value.split(','), (el) => {
          if (isNaN(el)) {
            const newValue = selectOptions[selectOptions.length-1].value + 1;
            selectOptions.push({
              label: el,
              value: newValue
            });
            return newValue;
          } else {
            return parseInt(el);
          }
        })
      });
    } else {
      if (isNaN(value)) {
        const newValue = selectOptions[selectOptions.length-1].value + 1;
        selectOptions.push({
          label: value,
          value: newValue
        });
        this.setState({
          selected: [newValue]
        });
      } else {
        this.setState({
          selected: [value]
        })
      }
    }

  },

  triggerAtSwitchToggle() {
    this.setState({ switchState: !this.state.switchState,
      selectOptions: this.state.selectOptions == demBallot ? repubBallot : demBallot });
  },
});

export default UiSwitchSandbox;
