// liveramp-ui-toolkit/js/sandboxComponents/uiWizardSandbox.jsx
import UiWizard from '../components/uiWizard.jsx';
import UiCheckbox from '../components/uiCheckbox.jsx';
import UiAlert from '../components/uiAlert.jsx';

const UiWizardSandbox = React.createClass({
  getInitialState () {
    return {
      checked: false,
      showAlert: false
    }
  },

  toggleChecked() {
    this.setState({
      checked: !this.state.checked
    });
  },

  confirmExit () {
    this.setState({
      showAlert: true
    });
  },

  render() {
    const steps = [
      {
        name: 'Step 1',
        component: <div>This is step 1</div>
      },
      {
        name: 'Step 2',
        component: (<div>
          <UiCheckbox
            label={"Click To Proceed"}
            checked={this.state.checked}
            handleChange={this.toggleChecked}
          />
        </div>),
        canProceed: () => this.state.checked
      },
      {
        name: 'Step 3',
        component: <div>This is step 3</div>
      }
    ]

    return (
      <div>
        {
          this.state.showAlert ? <UiAlert
            type={this.state.alertType}
            message={'You have canceled the wizard!'}
            permanent={false}
            handleClose={ () => { this.setState({showAlert: false}); } } 
          /> : null
        }
        <UiWizard steps={steps} cancel={this.confirmExit}/>
      </div>
    )
  }

});

export default UiWizardSandbox;
