// liveramp-ui-toolkit/js/sandboxComponents/uiWizardControlsSandbox.jsx
import UiWizardControls from '../components/uiWizardControls.jsx'

const UiWizardControlsSandbox = React.createClass({
  getInitialState () {
    return {
      currentPage: 1
    }
  },
  render () {
    return (
      <div className="wizard-example">
        <div className="col-md-8 col-md-offset-2">
          <div className="wizard-content" style={ {paddingBottom: '120px'} }>
            <div style={ {fontSize: '20px', textAlign: 'center', margin: '20px'} }>
              current page: { this.state.currentPage }
            </div>
            <div style={ {backgroundColor: 'black', height: '1000px', width: '20px'} }>
            </div>
          </div>
          <UiWizardControls
            buttonAttribs={{
              left: {
                handleChange: (e) => this.setState({currentPage: this.state.currentPage - 1}),
                disabled: this.state.currentPage === 1
              },
              cancel: {
                handleChange: (e) => alert('bye bye')
              },
              right: {
                handleChange: (e) => this.setState({currentPage: this.state.currentPage + 1}),
                disabled: this.state.currentPage === this.props.numPages
              }
            }}
          />
        </div>
      </div>
    )
  }
});

export default UiWizardControlsSandbox;
